src/main/scala/org/spartanz/parserz/ParsersModule.scala

```diff
@@ -23,6 +30,11 @@ trait ParsersModule extends ExprModule {
     final def filter[E1 >: E](e: E1)(f: Expr[A]): Grammar[SI, SO, E1, A] =
       Filter[SI, SO, E1, A](self, e, f)
 
+    final def flatMap[SI1 <: SI, SO1 >: SO, E1 >: E, B](f: A => Grammar[SI1, SO1, E1, B])(
+      implicit en: Enumerable[A]
+    ): Grammar[SI1, SO1, E1, B] =
+      Select(self.map(_ == en.max, b => if (b) en.max else en.min), f(en.max), f(en.min))
```

Comment:
@jdegoes 

Hi John!

I tried to implement the `flatMap` design you showed me at ScalaIO. The implementation I highlighted is not finished yet (it is not going through the range of values at this point), but parser/printer are correct and it is enough to show the feel of it (there is a test case in this PR as well).

I have a conceptual question though: what is the use case where it is practical enough?

My concerns are:
a. applicability (what can really be enumerable?)
b. performance (going through all the possibilities of `A`)

The use case I was trying to address is
1. parse input until value of `A` can be produced
2. use `A` to parse the rest of the input conditionally and produce `B`

For example:
1. parse bytes until `Header` is produced
2. use `Header` to decide which parser to use to get `Body`

But I think it can be achieved with existing functionality (pseudo-code below for brevity):
1. `HeaderParser.parse("some text)` which yields `Header("some)` and the rest of input `"text"`
2. `getBodyParser(header).parse("text")` which yields `Body("text")` and consumes all the input

Of course automatic reversing of the above is not possible, hence my attempts on `flatMap`. 

Could you please give me any feedback regarding concerns above? Thanks!

Reply:
You are right this cannot be done, because we are in "iso" land. 

You need to switch to the more category theoretic notion of `bind`, which is a less restrictive `join` operation that can take `F[F[A]]` and yield `F[A]`. 

Unfortunately this implies that to use such a thing, the user must have `B <=> F[A]`, such that they can take a `F[B]` and get a `F[F[A]]` out of it, on which to call `join`. It also implies invertibility of `join`. 

Let me think and get back, for now I'd say, try something simpler if you need this (e.g. `select`).

Reply:
@jdegoes 
Thanks John!

Yes, the idea of `Select extends Grammar` works, but is rather limited as I described above. I will finish and PR it so we can have a more concrete base for discussion.

May be we can talk again during Functional Scala? Thank you!

Reply:
Yes, let's try there!